From f0f22fb168ee6770e18cb1ac7fd2ee3cf814dd49 Mon Sep 17 00:00:00 2001
From: Jeffrey Clark <dude@zaplabs.com>
Date: Sat, 25 Jul 2015 20:08:43 -0500
Subject: [PATCH 01/17] allow libs to be linked multiple times

genie filters duplicate links but this is a problem when linking against external static libraries like libsdl with mingw.
winmm needs to be linked again after sdl because of the way gcc ld works. This patch corrects the chicken-n-egg problem.

Signed-off-by: Jeffrey Clark <dude@zaplabs.com>
---
 3rdparty/genie/src/base/bake.lua    |  8 ++++----
 3rdparty/genie/src/base/project.lua |  2 +-
 3 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/3rdparty/genie/src/base/bake.lua b/3rdparty/genie/src/base/bake.lua
index ab50612..f5da62a 100644
--- a/3rdparty/genie/src/base/bake.lua
+++ b/3rdparty/genie/src/base/bake.lua
@@ -157,15 +157,15 @@
 --    The source object, containing the settings to added to the destination.
 --
 
-	local function mergefield(kind, dest, src)
+	local function mergefield(kind, dest, src, fieldname)
 		local tbl = dest or { }
 		if kind == "keyvalue" or kind == "keypath" then
 			for key, value in pairs(src) do
-				tbl[key] = mergefield("list", tbl[key], value)
+				tbl[key] = mergefield("list", tbl[key], value, fieldname)
 			end
 		else
 			for _, item in ipairs(src) do
-				if not tbl[item] then
+				if fieldname == "links" or not tbl[item] then
 					table.insert(tbl, item)
 					tbl[item] = item
 				end
@@ -197,7 +197,7 @@
 				local field = premake.fields[fieldname]
 				if field then
 					if type(value) == "table" then
-						dest[fieldname] = mergefield(field.kind, dest[fieldname], value)
+						dest[fieldname] = mergefield(field.kind, dest[fieldname], value, fieldname)
 						if src.removes then
 							removes = src.removes[fieldname]
 							if removes then
diff --git a/3rdparty/genie/src/base/project.lua b/3rdparty/genie/src/base/project.lua
index 5806757..0312067 100644
--- a/3rdparty/genie/src/base/project.lua
+++ b/3rdparty/genie/src/base/project.lua
@@ -380,7 +380,7 @@
 				if pathstyle == "windows" and part ~= "object" then
 					item = path.translate(item, "\\")
 				end
-				if not table.contains(result, item) then
+				if kind == "system" or not table.contains(result, item) then
 					table.insert(result, item)
 				end
 			end
-- 
2.8.0

